src/lib/x509/x509_ext.cpp

```diff
@@ -764,6 +768,51 @@ void CRL_Issuing_Distribution_Point::decode_inner(const std::vector<uint8_t>& bu
    BER_Decoder(buf).decode(m_distribution_point).verify_end();
 }
 
+void TNAuthList::Entry::encode_into(DER_Encoder&) const {
+   throw Botan::Not_Implemented("TNAuthList extension serialization is not supported");
+}
+
+void TNAuthList::Entry::decode_from(class BER_Decoder& ber) {
+   BER_Object obj = ber.get_next_object();
+
+   m_type = static_cast<Type>(obj.type_tag());
+
+   switch(m_type) {
+      case TN_ServiceProviderCode: {
+         ASN1_String spc_string;
+         BER_Decoder(obj).decode(spc_string);
+         m_data = std::move(spc_string);
+      } break;
+      case TN_TelephoneNumberRange: {
+         m_data = RangeContainer();
+         auto& range_items = std::get<RangeContainer>(m_data);
+         BER_Decoder list = BER_Decoder(obj).start_sequence();
+         while(list.more_items()) {
+            TelephoneNumberRange entry;
+            list.decode(entry.start);
+            list.decode(entry.count);
+            range_items.emplace_back(std::move(entry));
+         }
+         list.end_cons();
+      } break;
+      case TN_TelephoneNumber: {
+         ASN1_String one_string;
+         BER_Decoder(obj).decode(one_string);
+         m_data = std::move(one_string);
+      } break;
+      default:
+         throw Decoding_Error("unexpected entry type");
```

Comment:
Let's make this a bit more explicit to help out the poor user who encounters this one day

```
throw Decoding_Error(fmt("Unexpected TNEntry type {}", m_type));
```